pkg/storage/cacher.go

```diff
@@ -355,9 +357,24 @@ func (c *Cacher) Codec() runtime.Codec {
 func (c *Cacher) processEvent(event watchCacheEvent) {
 	c.Lock()
 	defer c.Unlock()
-	for _, watcher := range c.watchers {
-		watcher.add(event)
-	}
+
+	// prevent creating too many goroutines
+	burstChan := make(chan struct{}, burstWatchers)
+	// prevent sending events to a watcher out of order
+	wait := sync.WaitGroup{}
+
+	wait.Add(len(c.watchers))
+	for _, w := range c.watchers {
+		burstChan <- struct{}{}
+		go func(watcher *cacheWatcher) {
+			defer func() {
+				<-burstChan
+				wait.Done()
+			}()
+			watcher.add(event)
+		}(w)
+	}
+	wait.Wait()
```

Comment:
@wojtek-t This implementation is just one possible solution. Another possible solution is use a larger buffer channel here https://github.com/mqliang/kubernetes/blob/cache-event-in-parallel/pkg/storage/cacher.go#L487, and once the channel blocked, close the watcher. 

Etcd address the problem like this, see:

* use a buffer channel with capacity 100: https://github.com/coreos/etcd/blob/master/store/watcher_hub.go#L69
* once blocked, close the watcher: https://github.com/coreos/etcd/blob/master/store/watcher.go#L61-L72

I am not sure which solution is better. Do you have any ideas?